Throw if `searchValue` is a non-global RegExp

mathiasbynens opened the pull request:
One of the potential solutions to #16. (If we decide to go this route, we'd want to change `matchAll` accordingly.)

mathiasbynens committed:
Throw if `searchValue` is a non-global RegExp

mathiasbynens commented:
@ljharb Any opinions re: the `spec.html` changes?

mathiasbynens committed:
Add matchAll changes

mathiasbynens commented:
Per your LGTMs, I've added the corresponding `matchAll` changes as well.

ljharb commented on spec.html:
```diff
@@ -76,3 +80,25 @@ <h1>String.prototype.replaceAll ( _searchValue_, _replaceValue_ )</h1>
+      1. <ins>If _isRegExp_ is true, then</ins>
+        1. <ins>Let _flags_ be ? ToString(? Get(_regexp_, `"flags"`)).</ins>
+        1. <ins>If _flags_ does not contain `"g"`, throw a *TypeError* exception.</ins>
+      1. Let _matcher_ be ? GetMethod(_regexp_, @@matchAll).
```
it does seem kind of odd that a non RegExp can have any kind of non “all” behavior it wants, but an actual RegExp is forced to have the proper “all” behavior. Not sure anything can be done about that, but it is an inconsistency introduced by this change.

mathiasbynens replied:
I see what you're saying. Still, there's value in making built-ins work well / in unsurprising ways by default, even if for userland subclasses, this is up to the user.

anba replied:
If we only/mostly care about built-ins, I wonder if makes more sense to move this check into <https://tc39.es/ecma262/#sec-regexp-prototype-matchall>?

mathiasbynens replied:
Interesting. I don't have a strong preference tbh. Any opinions? @schuay @ljharb

schuay replied:
No strong opinion. One argument for the current spot is that the location is consistent between `matchAll` and `replaceAll`. There's no `@@replaceAll` so we could not consistently move both checks to a RegExp builtin. I'm fine with either way.

ljharb replied:
It seems like the motivation/value of moving it inside the symbol methods is so a subclass could *override* the behavior - ie, could lack the g flag but still be “all” (vice versa is already going to be possible). I don’t see any benefit in allowing regex subclasses to deviate from this rather strong guard that we’ll have decided on - without concrete use cases, i think here is probably a better spot.

ljharb commented on spec.html:
```diff
@@ -76,3 +80,25 @@ <h1>String.prototype.replaceAll ( _searchValue_, _replaceValue_ )</h1>
+    1. If _regexp_ is neither *undefined* nor *null*, then
+      1. <ins>Let _isRegExp_ be ? IsRegExp(_regexp_).</ins>
+      1. <ins>If _isRegExp_ is true, then</ins>
+        1. <ins>Let _flags_ be ? ToString(? Get(_regexp_, `"flags"`)).</ins>
```
If flags is undefined, this should be the empty string, not the string undefined.

ljharb commented on spec.html:
```diff
@@ -42,8 +42,12 @@ <h1>String.prototype.replaceAll ( _searchValue_, _replaceValue_ )</h1>
     1. If _searchValue_ is neither *undefined* nor *null*, then
+      1. Let _isRegExp_ be ? IsRegExp(_searchString_).
+      1. If _isRegExp_ is true, then
+        1. Let _flags_ be ? ToString(? Get(_searchValue_, `"flags"`)).
```
If flags is undefined, this should be the empty string, not the string undefined.

mathiasbynens replied:
How about this?

```
1. Let _flags_ be ? Get(_searchValue_, `"flags"`).
1. If Type(_flags_) is String, then
  1. If _flags_ does not contain `"g"`, throw a *TypeError* exception.
```

Let me know if you had something else in mind.

ljharb replied:
That wouldn’t match the RegExp constructor, which tostrings everything but undefined.

I expect something like:
```
1. Let _flags_ be ? Get(_searchValue_, *”flags”*).
1. If Type(_flags_) is not Undefined,
  1. If ? ToString(_flags_) does not contain *”g”*, throw a *TypeError* exception.
```

jridgewell replied:
`"undefined"` doesn't contain `g`, so the end effect is the same. I think this falls under the same error as https://github.com/tc39/ecma262/pull/1517.

ljharb replied:
Yes, that happens to be the case, but I think the explicitness is important - especially so as to not in any way repeat the mistake that https://github.com/tc39/ecma262/pull/1517 is attempting to make.

In other words, the committee/engines may not have felt it worth it to fix that particular bug, but that doesn't mean I want to see an avoidable bug occur ever again, in any proposal.

mathiasbynens replied:
Done in e796ca6b1b8b668e57534c2ee2bf71a986820d1f.

jridgewell replied:
What if the value is `null`? Or `3`? Why special case only `undefined` here to skip the check? This also gets us back into the same broken state if you're improperly subclassed a regex:

```js
const r = /t/;
Object.defineProperty(r, 'flags', { value: undefined });

'test'.replaceAll(r, ''); // => 'est'
```

ljharb replied:
Because that’s how the RegExp constructor already operates.

jridgewell replied:
So what?

`exec` will happily except `null`. So will `split`, `search`, and all the others. The constructor is the least important precedent to follow here.

ljharb replied:
I don't agree; I think that the constructor is the only precedent to follow.

 - the constructor [special-cases `undefined`](https://tc39.es/ecma262/#sec-regexpinitialize)
 - `exec` reads the original flags slot ([step 5](https://tc39.es/ecma262/#sec-regexpbuiltinexec)), and so doesn't apply here
 - `split` unconditionally [stringifies](https://tc39.es/ecma262/#sec-regexp.prototype-@@split) `.flags` in ES2015+, but prior, it read from the internal slot
 - [`match`](https://tc39.es/ecma262/#sec-regexp.prototype-@@match) and [`search`](https://tc39.es/ecma262/#sec-regexp.prototype-@@search) and [`replace`](https://tc39.es/ecma262/#sec-regexp.prototype-@@replace) all delegate to `exec`, see above
 - [`matchAll`](https://tc39.es/ecma262/#sec-regexp-prototype-matchall) matches split, but this was my mistake the committee isn't yet willing to fix - it should have matched the constructor

Thus, only `split` actually has this incorrect behavior prior to my involvement. I'm contending that it is a bug/omission/oversight - not intentional - that split doesn't match the constructor, and it's my mistake that matchAll doesn't. I'm not willing to see another split-like deviation enter the language.

mathiasbynens committed:
Explicitly check for `undefined`, matching the RegExp constructor

https://tc39.es/ecma262/#sec-regexp-pattern-flags
https://github.com/tc39/proposal-string-replaceall/pull/24
